pinot-query-planner/src/main/java/org/apache/calcite/rel/rules/PinotInnerJoinToDynamicFilterRule.java

```diff
@@ -0,0 +1,159 @@
+ *    [ Transform ]     [ Transform ]           [Dyn. Filter] <-----
+ *         |                  |                      |              \
+ *    [Proj/Filter]     [Proj/Filter]           [ Transform ]    [xChange]
+ *         |                  |                      |                \
+ *    [Table Scan ]     [Table Scan ]           [Proj/Filter]    [ Transform ]
+ *                                                   |                 |
+ *                                              [Table Scan ]    [Proj/Filter]
+ *                                                                     |
+ *                                                               [Table Scan ]
+ *
+ * <p> The next part to compile the Dynamic filter into the Proj/Filter operator happens in the runtime.
+ *
+ * <p> This rewrite is only useful if we can ensure that:
+ * <ul>
+ *   <li>new plan can leverage the indexes and fast filtering of the leaf-stage Pinot server processing logic.</li>
+ *   <li>
+ *     data sending over xChange should be relatively small comparing to data would've been selected out if the dynamic
+ *     filter is not applied. E.g. selectivity of the dynamic filter should out-weight the data transfer overhead.
+ *   </li>
+ *   <li>
+ *     since leaf-stage operator can only process a typical Pinot V1 engine chain-operator chain. This means the entire
+ *     right-table will be ship over and persist in memory before the dynamic filter can occur. memory foot print will
+ *     be high so this rule should be use carefully until we have cost-based optimization.
+ *   </li>
+ * </ul>
+ */
+public class PinotInnerJoinToDynamicFilterRule extends RelOptRule {
+  public static final PinotInnerJoinToDynamicFilterRule INSTANCE =
+      new PinotInnerJoinToDynamicFilterRule(PinotRuleUtils.PINOT_REL_FACTORY);
+
+  public PinotInnerJoinToDynamicFilterRule(RelBuilderFactory factory) {
+    super(operand(LogicalJoin.class, any()), factory, null);
+  }
+
+  @Override
+  public boolean matches(RelOptRuleCall call) {
+    if (call.rels.length < 1) {
+      return false;
+    }
+    PinotRelOptPlannerContext context = (PinotRelOptPlannerContext) call.getPlanner().getContext();
+    if (call.rel(0) instanceof Join) {
+      Join join = call.rel(0);
+      JoinInfo joinInfo = join.analyzeCondition();
+      RelNode left = join.getLeft() instanceof HepRelVertex ? ((HepRelVertex) join.getLeft()).getCurrentRel()
+          : join.getLeft();
+      RelNode right = join.getRight() instanceof HepRelVertex ? ((HepRelVertex) join.getRight()).getCurrentRel()
+          : join.getRight();
+      return join.getJoinType() == JoinRelType.INNER && CollectionUtils.isEmpty(joinInfo.nonEquiConditions)
+          && left instanceof LogicalExchange && right instanceof LogicalExchange
+          && noExchangeInSubtree(left.getInput(0)) && noExchangeInSubtree(right.getInput(0))
+          && context.getOptions().containsKey(PinotRelOptPlannerContext.USE_DYNAMIC_FILTER);
```

Comment:
I suspect setting this flag will be done in a follow-up PR? I was trying to think about when we can enable this - it's a bit tricky to figure that out!

Reply:
- setting this flag is a user turn-on behavior at least for the first version
- once we have cost-based optimization, we can make the call based on table stats